### PR TITLE
Persistence Refactoring

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
- Fixed racing conditions in the tests
- Removed dead code
- Use C# 6
- Added XML documentation

### CODE REVIEW

File: src/core/Akka.Persistence/Persistent.cs [hunk 1]
```diff
@@ -67,53 +69,40 @@ public interface IPersistentEnvelope
     /// </summary>
     internal sealed class NonPersistentMessage : IPersistentEnvelope
     {
-        /// <summary>
-        /// TBD
-        /// </summary>
-        /// <param name="payload">TBD</param>
-        /// <param name="sender">TBD</param>
         public NonPersistentMessage(object payload, IActorRef sender)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/Persistent.cs]: Internal code should not contains `TBD Xml comments`

File: src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs [hunk 2]
```diff
@@ -249,22 +248,6 @@ private void HandleDeleteMessagesTo(DeleteMessagesTo message)
                 }, _continuationOptions);
         }
 
-        private void HandleReadHighestSequenceNr(ReadHighestSequenceNr message)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs]: Dead code

File: src/core/Akka.Persistence/InternalExtensions.cs [hunk 3]
```diff
@@ -1,34 +0,0 @@
...
-    /// </summary>
-    internal static class InternalExtensions
-    {
-        /// <summary>
-        /// Enqueues provided <paramref name="message"/> at the beginning of the internal actor cell mailbox message queue.
-        /// Requires current actor to use unbounded deque-based message queue. It will fail otherwise.
-        /// </summary>
-        /// <param name="context">TBD</param>
-        /// <param name="message">TBD</param>
-        public static void EnqueueMessageFirst(this IActorContext context, object message)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/InternalExtensions.cs]: Dead code

File: src/core/Akka.Persistence/Internal/CollectionsExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System.Collections.Immutable;
+
+namespace Akka.Persistence.Internal
+{
+    internal static class CollectionsExtensions
+    {
+        public static ImmutableArray<T> AddFirst<T>(this ImmutableArray<T> @this, T item)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/Internal/CollectionsExtensions.cs]: Emulated LinkedList

File: src/core/Akka.Persistence/JournalProtocol.cs [hunk 5]
```diff
@@ -861,216 +766,12 @@ public bool Equals(ReplayMessagesFailure other)
...
-        {
-            return $"ReplayMessagesFailure<cause: {Cause}>";
-        }
-    }
-
-    /// <summary>
-    /// TBD
-    /// </summary>
-    [Serializable]
-    public sealed class ReadHighestSequenceNr : IJournalRequest, IEquatable<ReadHighestSequenceNr>
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/JournalProtocol.cs]: Dead code from pre 2.4 persistence

File: src/core/Akka.Persistence.Tests/PersistentActorSpec.Actors.cs [hunk 6]
```diff
@@ -121,7 +123,7 @@ public override string ToString()
 
         internal abstract class ExamplePersistentActor : NamedPersistentActor
         {
-            protected LinkedList<object> Events = new LinkedList<object>();
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence.Tests/PersistentActorSpec.Actors.cs]: Protection from the racing conditions

File: src/core/Akka.Persistence.Tests/PersistentActorSpec.Actors.cs [hunk 7]
```diff
@@ -121,7 +123,7 @@ public override string ToString()
...
+            protected ImmutableArray<object> Events = ImmutableArray<object>.Empty;
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence.Tests/PersistentActorSpec.Actors.cs]: Protection from the racing conditions

File: src/core/Akka.Persistence/Internal/CollectionsExtensions.cs (hunk 4 above)
Comment (Horusiath) [inline src/core/Akka.Persistence/Internal/CollectionsExtensions.cs]: Have you checked the performance? AFAIK LinkedList AddFirst operation is O(1). Using ImmutableArray is copying the whole array into new struct: [implementation](https://github.com/dotnet/corefx/blob/29d7b6dee194e71bc5c98eec962bcac39bee8bb7/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs#L298).

**Comment by Horusiath** [comment]
Only one question left in comments. Beside that everything is fine.

File: src/core/Akka.Persistence/Internal/CollectionsExtensions.cs (hunk 4 above)
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/Internal/CollectionsExtensions.cs]: Yes, this one is slower. But I'm using it only in unit tests

File: src/core/Akka.Persistence/AtLeastOnceDeliverySemantic.cs [hunk 8]
```diff
@@ -260,37 +241,9 @@ protected MaxUnconfirmedMessagesExceededException(SerializationInfo info, Stream
...
-            public readonly DateTime Timestamp;
-
-            /// <summary>
-            /// TBD
-            /// </summary>
-            /// <param name="destination">TBD</param>
-            /// <param name="message">TBD</param>
-            /// <param name="timestamp">TBD</param>
-            /// <param name="attempt">TBD</param>
             public Delivery(ActorPath destination, object message, DateTime timestamp, int attempt)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Persistence/AtLeastOnceDeliverySemantic.cs]: This one will be internal in the next version, no need to have TBD docs

**Pull request merged**